### ISSUE TITLE
Is the combination of pipeline reasonable in the join scenario?

### ISSUE THREAD

**Issue by 744570676** [ClickHouse/ClickHouse#23105]
When i execute 'explain plan' command to find the combination of pipeline in the multiple join scenario, the following result is shown : 

```
Expression (Projection)                                                          
  Expression (Before ORDER BY and SELECT)                                        
    CreatingSets (Create sets for subqueries and joins)                          
      Aggregating                                                                
        Expression (Before GROUP BY)                                             
          Filter (WHERE)                                                         
            InflatingExpression (JOIN)                                           
              Expression (Before JOIN)                                           
                Expression (Projection)                                          
                  CreatingSets (Create sets for subqueries and joins)            
                    Expression (Before ORDER BY and SELECT)                      
                      Filter (WHERE)                                             
                        InflatingExpression (JOIN)                               
                          Expression (Before JOIN)                               
                            Expression (Projection)                              
                              CreatingSets (Create sets for subqueries and joins)
                                Expression (Before ORDER BY and SELECT)          
                                  Filter (WHERE)                                 
                                    InflatingExpression (JOIN)                   
                                      Expression (Before JOIN)                   
                                        ReadFromStorage (Read from MergeTree)
```
1. Isn't it more reasonable to put `InfaltingExpression (JOIN)` after `Fliter (where)`, so that the amount of data will be reduced a lot when joining？
2. CreatingSets read right table data using a LazyBlockInputStream to build the hash table，and InflatingExpression perform prebo operation. So, it's not easy to understand why put InfaltExpression before CreatingSets, i want to know how the pipeline is scheduled？
3. CreatingSets will resize the parallelism of current pipe to 1, which result in the poor join performance. What is the purpose of this?

**Comment by filimonov** [comment]
Some optimizations on the pipeline level were merged couple a of days ago, see https://github.com/ClickHouse/ClickHouse/pull/22087

Not sure if it addresses your issue, but it's worth checking the behavior of the same query on the latest testing release.

@KochetovNicolai do you want to add some comments here?

**Comment by KochetovNicolai** [comment]
Yes.
1. Is indeed partly implemented in #22087, but only for `LEFT` and `INNER` join.
2. It is not very logical indeed. Some legacy stuff. It would be more logical to have two inputs for join step.
3. Now it should not. I fixed in in some patch. But `CreatingSets` is still executed before join, so main query will not start until we  read right table.

**Comment by 744570676** [comment]
@filimonov @KochetovNicolai   thank you for your answer. The above result is in version v20.8.9.6-stable, so i try it again in the latest version v21.4.3.21-stable, the query plan is as follows:

```
Expression ((Projection + Before ORDER BY))                                                              
  CreatingSets (Create sets before main query execution)                                                 
    Aggregating                                                                                          
      Expression (Before GROUP BY)                                                                       
        Filter (WHERE)                                                                                   
          Join (JOIN)                                                                                    
            Expression ((Before JOIN + Projection))                                                      
              CreatingSets (Create sets before main query execution)                                     
                Expression (Before ORDER BY)                                                             
                  Filter (WHERE)                                                                         
                    Join (JOIN)                                                                          
                      Expression ((Before JOIN + Projection))                                            
                        CreatingSets (Create sets before main query execution)                           
                          Expression (Before ORDER BY)                                                   
                            Filter (WHERE)                                                               
                              Join (JOIN)                                                                
                                Expression (Before JOIN)                                                 
                                  SettingQuotaAndLimits (Set limits and quota after reading from storage)
                                    ReadFromPreparedSource (Read from NullSource)                        
                          CreatingSet (Create set for JOIN)                                              
                            Expression ((Projection + Before ORDER BY))                                  
                              Filter (WHERE)                                                             
                                SettingQuotaAndLimits (Set limits and quota after reading from storage)  
                                  ReadFromPreparedSource (Read from NullSource)                          
                CreatingSet (Create set for JOIN)                                                        
                  Expression ((Projection + Before ORDER BY))                                            
                    Filter (WHERE)                                                                       
                      SettingQuotaAndLimits (Set limits and quota after reading from storage)            
                        ReadFromPreparedSource (Read from NullSource)                                    
    CreatingSet (Create set for JOIN)                                                                    
      Expression ((Projection + Before ORDER BY))                                                        
        Filter (WHERE)                                                                                   
          SettingQuotaAndLimits (Set limits and quota after reading from storage)                        
            ReadFromPreparedSource (Read from NullSource)                                                
```

1. The query plan presents a tree structure. It can be easily seen that `Filter` is executed before `CreatingSet` during the `build pharse` of hash join, but `Filter` is still executed after `Join` during the `probe pharse` of hash join , which can be optimized later.

2. It would be better to implement a special processor `JoinTransform` to concat the pipelines of right and left tables , which has the higher performance than using `LazyBlockInputStream` due to the reduce of schedule and wait time.

3. I found the change in the source code. ^-^
